src/compiler/checker.ts

```diff
@@ -9613,6 +9614,9 @@ namespace ts {
                             context.inferredTypes[index] = inferredType = instantiatedConstraint;
                         }
                     }
+                    if (context.useAnyForNoInferences && !inferences.length && inferredType === emptyObjectType) {

```

Comment:
This means you never get the constraint if one exist, always get `any`, since any will always be assignable to any constraint. For `foo<T extends number>(): T`  `number` seemed like a better result than `any`.